File: apps/mobile/ios/HealthCoachAI/Views/Onboarding/OnboardingCoordinator.swift
import SwiftUI
import Foundation

// MARK: - Onboarding Step Enum
enum OnboardingStep: Int, CaseIterable {
    case splash = 0
    case authentication = 1
    case consent = 2
    case basicInfo = 3
    case lifestyle = 4
    case health = 5
    case preferences = 6
    case goals = 7
    case complete = 8
    
    var title: String {
        switch self {
        case .splash:
            return "Welcome"
        case .authentication:
            return "Sign In"
        case .consent:
            return "Privacy"
        case .basicInfo:
            return "Basic Info"
        case .lifestyle:
            return "Lifestyle"
        case .health:
            return "Health"
        case .preferences:
            return "Food Preferences"
        case .goals:
            return "Goals"
        case .complete:
            return "Complete"
        }
    }
    
    var description: String {
        switch self {
        case .splash:
            return "Let's start your health journey"
        case .authentication:
            return "Secure access to your account"
        case .consent:
            return "Your privacy matters to us"
        case .basicInfo:
            return "Tell us about yourself"
        case .lifestyle:
            return "Your daily habits"
        case .health:
            return "Health information"
        case .preferences:
            return "Food choices & preferences"
        case .goals:
            return "What you want to achieve"
        case .complete:
            return "Ready to start!"
        }
    }
}

// MARK: - Onboarding Data Models
struct OnboardingData {
    // Basic Info
    var firstName: String = ""
    var lastName: String = ""
    var email: String = ""
    var birthday: Date = Date()
    var gender: String = ""
    var height: Double = 170.0
    var weight: Double = 70.0
    var city: String = ""
    var state: String = ""
    
    // Lifestyle
    var activityLevel: String = "moderately_active"
    var smokingFrequency: Int = 0
    var alcoholFrequency: Int = 0
    var sleepHours: Double = 8.0
    var jobActivityLevel: Int = 3
    var eatingOutFrequency: Int = 3
    var stressLevel: Int = 5
    var waterIntake: Double = 2.5
    
    // Health
    var healthConditions: [String] = []
    var bloodPressureSystolic: Int = 120
    var bloodPressureDiastolic: Int = 80
    var fastingBloodSugar: Int = 90
    var emergencyContactName: String = ""
    var emergencyContactPhone: String = ""
    
    // Preferences
    var dietaryPreference: String = "vegetarian"
    var favoriteCuisines: [String] = []
    var allergens: [String] = []
    var spiceTolerance: String = "medium"
    var favoriteIngredients: [String] = []
    var dislikedIngredients: [String] = []
    var cravings: [String] = []
    var mealsPerDay: Int = 3
    var cookingSkillLevel: Int = 3
    
    // Goals
    var primaryGoal: String = "weight_loss"
    var goalPriority: String = "medium"
    var intensity: String = "moderate"
    var targetWeight: Double = 65.0
    var targetDate: Date = Calendar.current.date(byAdding: .month, value: 6, to: Date()) ?? Date()
    var dailyCalorieTarget: Int = 1800
    var weeklyExerciseTarget: Int = 150
    var motivation: String = ""
}

// MARK: - Onboarding Coordinator
class OnboardingCoordinator: ObservableObject {
    @Published var currentStep: OnboardingStep = .splash
    @Published var isOnboardingComplete: Bool = false
    @Published var onboardingData = OnboardingData()
    @Published var progress: Double = 0.0
    @Published var isLoading: Bool = false
    @Published var errorMessage: String = ""
    
    private let totalSteps = OnboardingStep.allCases.count - 1 // Exclude complete step
    
    init() {
        updateProgress()
        checkOnboardingStatus()
    }
    
    // MARK: - Navigation
    func nextStep() {
        guard currentStep.rawValue < OnboardingStep.complete.rawValue else {
            completeOnboarding()
            return
        }
        
        if let nextStep = OnboardingStep(rawValue: currentStep.rawValue + 1) {
            withAnimation(.easeInOut(duration: 0.3)) {
                currentStep = nextStep
                updateProgress()
            }
        }
    }
    
    func previousStep() {
        guard currentStep.rawValue > OnboardingStep.splash.rawValue else { return }
        
        if let previousStep = OnboardingStep(rawValue: currentStep.rawValue - 1) {
            withAnimation(.easeInOut(duration: 0.3)) {
                currentStep = previousStep
                updateProgress()
            }
        }
    }
    
    func skipStep() {
        // Log analytics event for skipped step
        AnalyticsManager.shared.trackEvent("onboarding_step_skipped", properties: [
            "step": currentStep.rawValue,
            "step_name": currentStep.title
        ])
        
        nextStep()
    }
    
    func goToStep(_ step: OnboardingStep) {
        withAnimation(.easeInOut(duration: 0.3)) {
            currentStep = step
            updateProgress()
        }
    }
    
    // MARK: - Data Management
    func saveBasicInfo() async {
        isLoading = true
        errorMessage = ""
        
        do {
            let apiData = OnboardingBasicInfoAPI(
                firstName: onboardingData.firstName,
                lastName: onboardingData.lastName,
                email: onboardingData.email.isEmpty ? nil : onboardingData.email,
                birthday: ISO8601DateFormatter().string(from: onboardingData.birthday),
                gender: onboardingData.gender,
                height: onboardingData.height,
                weight: onboardingData.weight,
                city: onboardingData.city.isEmpty ? nil : onboardingData.city,
                state: onboardingData.state.isEmpty ? nil : onboardingData.state
            )
            
            let result = try await OnboardingAPIService.shared.saveBasicInfo(apiData)
            
            if result.success {
                AnalyticsManager.shared.trackEvent("onboarding_basic_info_completed", properties: [
                    "step": currentStep.rawValue
                ])
                nextStep()
            } else {
                errorMessage = "Failed to save basic information. Please try again."
            }
        } catch {
            errorMessage = "Network error. Please check your connection and try again."
        }
        
        isLoading = false
    }
    
    func saveLifestyle() async {
        isLoading = true
        errorMessage = ""
        
        do {
            let apiData = OnboardingLifestyleAPI(
                activityLevel: onboardingData.activityLevel,
                smokingFrequency: onboardingData.smokingFrequency,
                alcoholFrequency: onboardingData.alcoholFrequency,
                sleepHours: onboardingData.sleepHours,
                jobActivityLevel: onboardingData.jobActivityLevel,
                eatingOutFrequency: onboardingData.eatingOutFrequency,
                stressLevel: onboardingData.stressLevel,
                waterIntake: onboardingData.waterIntake
            )
            
            let result = try await OnboardingAPIService.shared.saveLifestyle(apiData)
            
            if result.success {
                AnalyticsManager.shared.trackEvent("onboarding_lifestyle_completed", properties: [
                    "step": currentStep.rawValue,
                    "activity_level": onboardingData.activityLevel
                ])
                nextStep()
            } else {
                errorMessage = "Failed to save lifestyle information. Please try again."
            }
        } catch {
            errorMessage = "Network error. Please check your connection and try again."
        }
        
        isLoading = false
    }
    
    func saveHealth() async {
        isLoading = true
        errorMessage = ""
        
        do {
            let apiData = OnboardingHealthAPI(
                healthConditions: onboardingData.healthConditions,
                bloodPressureSystolic: onboardingData.bloodPressureSystolic,
                bloodPressureDiastolic: onboardingData.bloodPressureDiastolic,
                fastingBloodSugar: onboardingData.fastingBloodSugar,
                emergencyContactName: onboardingData.emergencyContactName.isEmpty ? nil : onboardingData.emergencyContactName,
                emergencyContactPhone: onboardingData.emergencyContactPhone.isEmpty ? nil : onboardingData.emergencyContactPhone
            )
            
            let result = try await OnboardingAPIService.shared.saveHealth(apiData)
            
            if result.success {
                AnalyticsManager.shared.trackEvent("onboarding_health_completed", properties: [
                    "step": currentStep.rawValue,
                    "has_health_conditions": !onboardingData.healthConditions.isEmpty
                ])
                nextStep()
            } else {
                errorMessage = "Failed to save health information. Please try again."
            }
        } catch {
            errorMessage = "Network error. Please check your connection and try again."
        }
        
        isLoading = false
    }
    
    func savePreferences() async {
        isLoading = true
        errorMessage = ""
        
        do {
            let apiData = OnboardingPreferencesAPI(
                dietaryPreference: onboardingData.dietaryPreference,
                favoriteCuisines: onboardingData.favoriteCuisines,
                allergens: onboardingData.allergens,
                spiceTolerance: onboardingData.spiceTolerance,
                favoriteIngredients: onboardingData.favoriteIngredients,
                dislikedIngredients: onboardingData.dislikedIngredients,
                cravings: onboardingData.cravings,
                mealsPerDay: onboardingData.mealsPerDay,
                cookingSkillLevel: onboardingData.cookingSkillLevel
            )
            
            let result = try await OnboardingAPIService.shared.savePreferences(apiData)
            
            if result.success {
                AnalyticsManager.shared.trackEvent("onboarding_preferences_completed", properties: [
                    "step": currentStep.rawValue,
                    "dietary_preference": onboardingData.dietaryPreference
                ])
                nextStep()
            } else {
                errorMessage = "Failed to save preferences. Please try again."
            }
        } catch {
            errorMessage = "Network error. Please check your connection and try again."
        }
        
        isLoading = false
    }
    
    func saveGoals() async {
        isLoading = true
        errorMessage = ""
        
        do {
            let apiData = OnboardingGoalsAPI(
                primaryGoal: onboardingData.primaryGoal,
                goalPriority: onboardingData.goalPriority,
                intensity: onboardingData.intensity,
                targetWeight: onboardingData.targetWeight,
                targetDate: ISO8601DateFormatter().string(from: onboardingData.targetDate),
                dailyCalorieTarget: onboardingData.dailyCalorieTarget,
                weeklyExerciseTarget: onboardingData.weeklyExerciseTarget,
                motivation: onboardingData.motivation.isEmpty ? nil : onboardingData.motivation
            )
            
            let result = try await OnboardingAPIService.shared.saveGoals(apiData)
            
            if result.onboardingComplete {
                AnalyticsManager.shared.trackEvent("onboarding_completed", properties: [
                    "primary_goal": onboardingData.primaryGoal,
                    "completed_at": ISO8601DateFormatter().string(from: Date())
                ])
                completeOnboarding()
            } else {
                errorMessage = "Failed to save goals. Please try again."
            }
        } catch {
            errorMessage = "Network error. Please check your connection and try again."
        }
        
        isLoading = false
    }
    
    // MARK: - Private Methods
    private func updateProgress() {
        progress = Double(currentStep.rawValue) / Double(totalSteps)
    }
    
    private func completeOnboarding() {
        withAnimation(.easeInOut(duration: 0.5)) {
            currentStep = .complete
            progress = 1.0
        }
        
        // Save completion status to UserDefaults
        UserDefaults.standard.set(true, forKey: "onboardingCompleted")
        
        // Delay before showing main app
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            withAnimation(.easeInOut(duration: 0.5)) {
                self.isOnboardingComplete = true
            }
        }
    }
    
    private func checkOnboardingStatus() {
        // Check if user has already completed onboarding
        let isCompleted = UserDefaults.standard.bool(forKey: "onboardingCompleted")
        if isCompleted {
            isOnboardingComplete = true
        }
        
        // TODO: Also check with backend API for onboarding progress
        // This will help handle cases where user completed onboarding on another device
    }
}

// MARK: - Analytics Helper
class AnalyticsManager {
    static let shared = AnalyticsManager()
    
    private init() {}
    
    func trackEvent(_ eventName: String, properties: [String: Any]? = nil) {
        // Privacy-safe analytics tracking
        // Only track flow events, no PII
        print("📊 Analytics: \(eventName) - \(properties ?? [:])")
        
        // TODO: Integrate with analytics service (Firebase, Mixpanel, etc.)
        // Ensure compliance with privacy policies
    }
}

File: apps/mobile/android/app/src/main/java/com/healthcoachai/app/onboarding/ui/OnboardingCoordinator.kt
package com.healthcoachai.app.onboarding.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.healthcoachai.app.onboarding.data.OnboardingRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject
import java.util.Date

/**
 * OnboardingCoordinator - Manages the onboarding flow state and navigation
 * Implements Phase 8 requirements for guided onboarding with conversational UX
 */
@HiltViewModel
class OnboardingCoordinator @Inject constructor(
    private val onboardingRepository: OnboardingRepository
) : ViewModel() {
    
    // MARK: - State
    private val _currentStep = MutableStateFlow(OnboardingStep.SPLASH)
    val currentStep: StateFlow<OnboardingStep> = _currentStep.asStateFlow()
    
    private val _isOnboardingComplete = MutableStateFlow(false)
    val isOnboardingComplete: StateFlow<Boolean> = _isOnboardingComplete.asStateFlow()
    
    private val _onboardingData = MutableStateFlow(OnboardingData())
    val onboardingData: StateFlow<OnboardingData> = _onboardingData.asStateFlow()
    
    private val _progress = MutableStateFlow(0f)
    val progress: StateFlow<Float> = _progress.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow("")
    val errorMessage: StateFlow<String> = _errorMessage.asStateFlow()
    
    private val totalSteps = OnboardingStep.values().size - 1 // Exclude COMPLETE
    
    init {
        updateProgress()
        checkOnboardingStatus()
    }
    
    // MARK: - Navigation
    fun nextStep() {
        val currentStepValue = _currentStep.value
        if (currentStepValue.ordinal < OnboardingStep.COMPLETE.ordinal) {
            val nextStep = OnboardingStep.values()[currentStepValue.ordinal + 1]
            _currentStep.value = nextStep
            updateProgress()
        } else {
            completeOnboarding()
        }
    }
    
    fun previousStep() {
        val currentStepValue = _currentStep.value
        if (currentStepValue.ordinal > OnboardingStep.SPLASH.ordinal) {
            val previousStep = OnboardingStep.values()[currentStepValue.ordinal - 1]
            _currentStep.value = previousStep
            updateProgress()
        }
    }
    
    fun skipStep() {
        // Log analytics event for skipped step
        trackEvent("onboarding_step_skipped", mapOf(
            "step" to _currentStep.value.ordinal,
            "step_name" to _currentStep.value.name
        ))
        
        nextStep()
    }
    
    fun goToStep(step: OnboardingStep) {
        _currentStep.value = step
        updateProgress()
    }
    
    // MARK: - Data Management
    fun saveBasicInfo() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = ""
            
            try {
                val result = onboardingRepository.saveBasicInfo(onboardingData.value)
                
                if (result.isSuccess) {
                    trackEvent("onboarding_basic_info_completed", mapOf(
                        "step" to _currentStep.value.ordinal
                    ))
                    nextStep()
                } else {
                    _errorMessage.value = "Failed to save basic information. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to save basic information. Please try again."
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun saveLifestyle() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = ""
            
            try {
                val result = onboardingRepository.saveLifestyle(onboardingData.value)
                
                if (result.isSuccess) {
                    trackEvent("onboarding_lifestyle_completed", mapOf(
                        "step" to _currentStep.value.ordinal,
                        "activity_level" to onboardingData.value.activityLevel
                    ))
                    nextStep()
                } else {
                    _errorMessage.value = "Failed to save lifestyle information. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to save lifestyle information. Please try again."
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun saveHealth() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = ""
            
            try {
                val result = onboardingRepository.saveHealth(onboardingData.value)
                
                if (result.isSuccess) {
                    trackEvent("onboarding_health_completed", mapOf(
                        "step" to _currentStep.value.ordinal,
                        "has_health_conditions" to onboardingData.value.healthConditions.isNotEmpty()
                    ))
                    nextStep()
                } else {
                    _errorMessage.value = "Failed to save health information. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to save health information. Please try again."
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun savePreferences() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = ""
            
            try {
                val result = onboardingRepository.savePreferences(onboardingData.value)
                
                if (result.isSuccess) {
                    trackEvent("onboarding_preferences_completed", mapOf(
                        "step" to _currentStep.value.ordinal,
                        "dietary_preference" to onboardingData.value.dietaryPreference
                    ))
                    nextStep()
                } else {
                    _errorMessage.value = "Failed to save preferences. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to save preferences. Please try again."
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun saveGoals() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = ""
            
            try {
                val result = onboardingRepository.saveGoals(onboardingData.value)
                
                if (result.isSuccess) {
                    trackEvent("onboarding_completed", mapOf(
                        "primary_goal" to onboardingData.value.primaryGoal,
                        "completed_at" to Date().toString()
                    ))
                    completeOnboarding()
                } else {
                    _errorMessage.value = "Failed to save goals. Please try again."
                }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to save goals. Please try again."
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    fun updateBasicInfo(update: (OnboardingData) -> OnboardingData) {
        _onboardingData.value = update(_onboardingData.value)
    }
    
    fun updateLifestyle(update: (OnboardingData) -> OnboardingData) {
        _onboardingData.value = update(_onboardingData.value)
    }
    
    fun updateHealth(update: (OnboardingData) -> OnboardingData) {
        _onboardingData.value = update(_onboardingData.value)
    }
    
    fun updatePreferences(update: (OnboardingData) -> OnboardingData) {
        _onboardingData.value = update(_onboardingData.value)
    }
    
    fun updateGoals(update: (OnboardingData) -> OnboardingData) {
        _onboardingData.value = update(_onboardingData.value)
    }
    
    // MARK: - Private Methods
    private fun updateProgress() {
        _progress.value = _currentStep.value.ordinal.toFloat() / totalSteps.toFloat()
    }
    
    private fun completeOnboarding() {
        _currentStep.value = OnboardingStep.COMPLETE
        _progress.value = 1.0f
        
        // Save completion status to local storage
        viewModelScope.launch {
            try {
                // Save locally (implementation pending - to be added with SharedPreferences)
                // SharedPreferences: isOnboardingCompleted = true
                
                // Delay before showing main app
                kotlinx.coroutines.delay(2000)
                _isOnboardingComplete.value = true
            } catch (e: Exception) {
                // Handle error silently, onboarding is functionally complete
                _isOnboardingComplete.value = true
            }
        }
    }
    
    private fun checkOnboardingStatus() {
        // Check if user has already completed onboarding
        viewModelScope.launch {
            try {
                val result = onboardingRepository.checkOnboardingStatus()
                if (result.isSuccess && result.getOrDefault(false)) {
                    _isOnboardingComplete.value = true
                }
                // If false or error, keep showing onboarding
            } catch (e: Exception) {
                // On error, show onboarding to be safe
                // This ensures users can always access the app
            }
        }
    }
    
    private fun trackEvent(eventName: String, properties: Map<String, Any>) {
        // Privacy-safe analytics tracking
        // Only track flow events, no PII
        println("📊 Analytics: $eventName - $properties")
        
        // Analytics integration (to be implemented with proper service)
        // Ensure compliance with privacy policies and user consent
        // Example: Firebase Analytics, Mixpanel, or custom analytics
        viewModelScope.launch {
            try {
                // analyticsService.track(eventName, properties)
            } catch (e: Exception) {
                // Analytics failures should not affect user experience
                println("Analytics error: ${e.message}")
            }
        }
    }
}

// MARK: - Data Models
enum class OnboardingStep(val title: String, val description: String) {
    SPLASH("Welcome", "Let's start your health journey"),
    AUTHENTICATION("Sign In", "Secure access to your account"),
    CONSENT("Privacy", "Your privacy matters to us"),
    BASIC_INFO("Basic Info", "Tell us about yourself"),
    LIFESTYLE("Lifestyle", "Your daily habits"),
    HEALTH("Health", "Health information"),
    PREFERENCES("Food Preferences", "Food choices & preferences"),
    GOALS("Goals", "What you want to achieve"),
    COMPLETE("Complete", "Ready to start!")
}

data class OnboardingData(
    // Basic Info
    val firstName: String = "",
    val lastName: String = "",
    val email: String = "",
    val birthday: Date = Date(),
    val gender: String = "",
    val height: Double = 170.0,
    val weight: Double = 70.0,
    val city: String = "",
    val state: String = "",
    
    // Lifestyle
    val activityLevel: String = "moderately_active",
    val smokingFrequency: Int = 0,
    val alcoholFrequency: Int = 0,
    val sleepHours: Double = 8.0,
    val jobActivityLevel: Int = 3,
    val eatingOutFrequency: Int = 3,
    val stressLevel: Int = 5,
    val waterIntake: Double = 2.5,
    
    // Health
    val healthConditions: List<String> = emptyList(),
    val bloodPressureSystolic: Int = 120,
    val bloodPressureDiastolic: Int = 80,
    val fastingBloodSugar: Int = 90,
    val emergencyContactName: String = "",
    val emergencyContactPhone: String = "",
    
    // Preferences
    val dietaryPreference: String = "vegetarian",
    val favoriteCuisines: List<String> = emptyList(),
    val allergens: List<String> = emptyList(),
    val spiceTolerance: String = "medium",
    val favoriteIngredients: List<String> = emptyList(),
    val dislikedIngredients: List<String> = emptyList(),
    val cravings: List<String> = emptyList(),
    val mealsPerDay: Int = 3,
    val cookingSkillLevel: Int = 3,
    
    // Goals
    val primaryGoal: String = "weight_loss",
    val goalPriority: String = "medium",
    val intensity: String = "moderate",
    val targetWeight: Double = 65.0,
    val targetDate: Date = Date(),
    val dailyCalorieTarget: Int = 1800,
    val weeklyExerciseTarget: Int = 150,
    val motivation: String = ""
)